File: sem1/hw5/5.5/stack.cpp
#include <iostream>
#include "stack.h"
using namespace std;

Stack * createStack()
{
    return new Stack {nullptr};
}

void push(Stack *stack, int value)
{
    StackElement *newElement = new StackElement {value, stack->first};
    stack->first = newElement;
}

int pop(Stack *stack)
{
    if (!isEmpty(stack))
    {
        int result = stack->first->value;
        StackElement *newElement = stack->first->next;
        delete stack->first;
        stack->first = newElement;
        return result;
    }
    return -1;
}

bool isEmpty(Stack *stack)
{
    return stack->first == nullptr;
}

void deleteStack(Stack *stack)
{
    StackElement *current = stack->first;
    while (current)
    {
        StackElement *newElement = current->next;
        delete current;
        current = newElement;
    }
    delete current;
    delete stack;
}

File: sem2/hw1/1.1/Stack.java
package com.AntonChernikov.g144;

public class Stack {
    private class StackElement {
        private int value;
        private StackElement next;

        private StackElement(int value, StackElement next) {
            this.value = value;
            this.next = next;
        }
    }

    private StackElement first = null;
    public int size = 0;

    public void push(int value) {
        StackElement newElement = new StackElement(value, first);
        first = newElement;
        size++;
    }

    public int pop() {
        if (first != null) {
            int result = first.value;
            first = first.next;
            size--;
            return result;
        }
        return 0;
    }

    public boolean isEmpty() {
        return first == null;
    }

    public int size() {
        return size;
    }

    public void print() {
        if (this.isEmpty()) {
            System.out.println("Stack is empty");
        }
        else {
            StackElement current = first;
            while (current != null) {
                System.out.print(current.value + " ");
                current = current.next;
            }
        }
        System.out.println();
    }
}
